src/Analysis/Engine/Impl/Indexing/IndexerWalker.cs

```diff
@@ -0,0 +1,245 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using Microsoft.PythonTools.Parsing.Ast;
+
+namespace Microsoft.PythonTools.Analysis.Indexing {
+    internal class IndexerWalker : PythonWalker {
+        private readonly PythonAst _ast;
+
+        private readonly Stack<HierarchicalSymbol> _scope = new Stack<HierarchicalSymbol>();
+        private HierarchicalSymbol Parent => _scope.Count > 0 ? _scope.Peek() : null;
+
+        private bool _skipNext;
+        private bool SkipNext {
```

Comment:
Property is a method that looks like field, so one of the expectation about the property in the single-threaded context is that sequential call produces the same result, i.e.:
```cs
walker.SkipNext = true;
var a = walker.SkipNext;
var b = walker.SkipNext;
Assert.Equal(a, b);
```

Reply:
That's a good point, I'll switch it to a function (or just access `_skipNext`). I think I might have had too much fun making a property with side effects.

Reply:
By the way, the only reason I have `SkipNext` is to skip over function/class NameExpressions, since I pick them up in their definitions instead. I do it like this because I don't seem to have a way to control how things are walked (i.e. it's the function definition node which controls how it is walked, something I can't override). Is there some obvious way to modify how it's walking, or is this a valid thing to do to get around the issue?

Reply:
You can walk the AST manually. See https://github.com/MikhailArkhipov/python-language-server/blob/analysis10/src/Analysis/Ast/Impl/Analyzer/Symbols/ClassEvaluator.cs#L68

Reply:
You can also override just `Expression` and handle cases in a switch.

Reply:
My rework walks manually when it's needed, instead of the `SkipNext` thing, so resolving.